File: 백준/Bronze/1978. 소수 찾기/소수 찾기.py
from sys import stdin

input = stdin.readline

_ = int(input().rstrip())
lst = list(map(int, input().split()))

# 에라토스테네스의 체
MAX = 1_000
sieve = [False, False] + [True] * (MAX - 1)
for i in range(2, int(MAX ** (1 / 2)) + 1):
    if sieve[i]:
        for j in range(i * i, MAX + 1, i):
            sieve[j] = False  # 배수는 소수가 아니다.

cnt = 0
for num in lst:
    if sieve[num]:
        cnt += 1

print(cnt)

File: Java/프로그래머스/2/42839. 소수 찾기/소수 찾기.java
import java.io.*;
import java.util.*;

class Solution {
    
    static int MAX = 10_000_000;
    static int N; 
    static boolean[] visited;
    static boolean[] sieve = new boolean[MAX + 1];
    static StringBuilder sb = new StringBuilder();
    static Set<Integer> set = new HashSet<>();
    
    public int solution(String numbers) {
        
        // 에라토스테네스의 체 구현
        sieve[0] = true;
        sieve[1] = true;
        for (int i = 2; i <= Math.sqrt(MAX); i++) {
            if (!sieve[i]) {
                for (int j = i * i; j <= MAX; j+=i)
                    sieve[j] = true; // 합성수 처리 
            }
        }
                
        N = numbers.length(); // 숫자의 개수 
        visited = new boolean[N];
        
        bt(0, new ArrayList<Integer>(), numbers.toCharArray());
        
        return set.size();
    }
    
    private static void bt(int cnt, ArrayList<Integer> lst, char[] nums) {
        
        if (!lst.isEmpty() && cnt <= N) {

            for (Integer idx : lst)
                sb.append(nums[idx]);

            int num = Integer.parseInt(sb.toString());
            if (!sieve[num]) {
                set.add(num);
            }

            sb.setLength(0);
            if (cnt == N)
                return;
        }
        
        for (int i = 0; i < N; i++) {
            if (!visited[i]) {
                visited[i] = true; // 방문 처리 
                lst.add(i);
                bt(cnt + 1, lst, nums);
                lst.remove(lst.size() - 1);
                visited[i] = false; // 복원 처리
            }
        }   
    }
}